server/src/test/java/org/elasticsearch/cluster/health/ClusterHealthAllocationTests.java

```diff
@@ -0,0 +1,102 @@
+        MetaData metaData = MetaData.builder()
+            .put(IndexMetaData.builder("test")
+                .settings(settings(Version.CURRENT))
+                .numberOfShards(2)
+                .numberOfReplicas(1))
+            .build();
+        RoutingTable routingTable = RoutingTable.builder().addAsNew(metaData.index("test")).build();
+        clusterState = ClusterState.builder(clusterState).metaData(metaData).routingTable(routingTable).build();
+        MockAllocationService allocation = createAllocationService();
+        clusterState = allocation.reroute(clusterState, "reroute");
+        assertTrue(clusterState.nodes().getDataNodes().size() == 0);
+        assertTrue(clusterState.nodes().getMasterNodes().size() == 1);
+        assertEquals(ClusterHealthStatus.RED, getClusterHealthStatus(clusterState));
+        clusterState = addNode(clusterState, "node_d1", false);
+        assertTrue(clusterState.nodes().getDataNodes().size() == 1);
+        clusterState = allocation.reroute(clusterState, "reroute");
+        // starting primaries
+        clusterState = allocation.applyStartedShards(clusterState, clusterState.getRoutingNodes().shardsWithState(INITIALIZING));
+        assertEquals(ClusterHealthStatus.YELLOW, getClusterHealthStatus(clusterState));
+        clusterState = addNode(clusterState, "node_d2", false);
+        clusterState = allocation.reroute(clusterState, "reroute");
+        // starting replicas
+        clusterState = allocation.applyStartedShards(clusterState, clusterState.getRoutingNodes().shardsWithState(INITIALIZING));
+        assertEquals(ClusterHealthStatus.GREEN, getClusterHealthStatus(clusterState));
+        clusterState = removeNode(clusterState, "node_d1");
+        assertEquals(ClusterHealthStatus.YELLOW, getClusterHealthStatus(clusterState));
+        clusterState = removeNode(clusterState, "node_d2");
+        assertEquals(ClusterHealthStatus.RED, getClusterHealthStatus(clusterState));
+        routingTable = RoutingTable.builder(routingTable).remove("test").build();
+        metaData = MetaData.builder(clusterState.metaData()).remove("test").build();
+        clusterState = ClusterState.builder(clusterState).routingTable(routingTable).metaData(metaData).build();
+        assertTrue(clusterState.nodes().getDataNodes().size() == 0);
+        assertTrue(clusterState.nodes().getMasterNodes().size() == 1);
+        assertEquals(ClusterHealthStatus.GREEN, getClusterHealthStatus(clusterState));
+    }
+
+    private ClusterState addNode(ClusterState clusterState, String nodeName, boolean isMaster) {
+
+        DiscoveryNodes.Builder nodeBuilder = DiscoveryNodes.builder(clusterState.getNodes());
+        if (isMaster) {
+            nodeBuilder = nodeBuilder.add(newNode(nodeName, Collections.singleton(DiscoveryNode.Role.MASTER)));
+        } else {
+            nodeBuilder = nodeBuilder.add(newNode(nodeName, Collections.singleton(DiscoveryNode.Role.DATA)));
```

Comment:
No need to mutate `nodeBuilder` here:

```suggestion
           nodeBuilder.add(newNode(nodeName, Collections.singleton(DiscoveryNode.Role.DATA)));
```